# 미션 - 숫자 야구

## 📝 목차

1. [🕹 게임 실행 예시](#게임-실행-예시)
2. [🛠 기능 목록](#기능-목록)
3. [🤔 개선이 필요한 부분](#개선이-필요한-부분)
4. [😊 배운 점](#배운-점)

## 게임 실행 예시

![야구게임](https://user-images.githubusercontent.com/33623078/200571464-4054b05b-a07f-4a2b-88b7-660bb9fc00d7.gif)

## 기능 목록

### ❗️ 게임 실행 흐름

- [x] 게임 시작
  - [x] 컴퓨터의 수 생성 (1에서 9까지 서로 다른 임의의 수 3개)
  - [x] 시작을 알리는 텍스트 `숫자 야구 게임을 시작합니다.`
- [x] 게임 반복
  - [x] 숫자 입력 요청 텍스트 `숫자를 입력해주세요 :` 출력 시 사용자 입력 받기
    - [x] 입력이 유효하지 않은 경우 예외처리
      - [x] 3자리의 수가 아닌 경우
      - [x] 각 자릿수가 1에서 9사이의 숫자가 아닌 경우
      - [x] 각 자릿수가 각각 서로 다른 숫자가 아닌 경우
    - [x] 입력한 숫자에 대한 결과 출력
      - [x] `n볼 m스트라이크`
      - [x] `n볼`
      - [x] `n스트라이크`
      - [x] 하나도 없는 경우 `낫싱`
  - [x] 정답 시 게임 재시작 텍스트 `게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.` 출력 후 사용자 입력 받기
    - [x] 입력이 유효하지 않은 경우 예외처리
      - [x] 1과 2 중 하나의 수가 아닌 경우
    - [x] 분기
      - [x] 1: 게임 처음부터 다시 시작
      - [x] 2: 게임 종료
- [x] 게임 종료
  - [x] 입력이 유효하지 않은 경우

### 사용자 입력 관련

- 유효성 검사
  - 게임 중
    - [x] 입력이 3자리의 수인가?
    - [x] 입력의 각 자릿수가 1에서 9사이의 숫자인가?
    - [x] 입력의 각 자릿 수가 각각 서로 다른 수인가?
  - 게임 종료 시
    - [x] 입력이 1과 2 중 하나의 수인가?
- 유효성 검사에 위배되는 경우
  - [x] `throw`문을 사용해 예외를 발생시킨다
  - [x] 예외 발생 시 사용자에게 이유를 출력으로 알려준다
  - [x] 애플리케이션 종료

### 컴퓨터 출력 관련

- 게임 시작할 때
  - [x] 시작을 알리는 텍스트 `숫자 야구 게임을 시작합니다.`
- 게임 진행중
  - [x] 숫자 입력 요청 텍스트 `숫자를 입력해주세요 :`
  - 입력한 숫자에 대한 결과 출력
    - [x] `n볼 m스트라이크`
    - [x] `n볼`
    - [x] `n스트라이크`
    - [x] 하나도 없는 경우 `낫싱`
  - 3개의 숫자를 모두 맞힌 경우 (3스트라이크)
    - [x] 게임 종료 텍스트 `3개의 숫자를 모두 맞히셨습니다! 게임 종료`
    - [x] 게임 재시작 텍스트`게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.`
- 입력 유효성 위배
  - [x] 3자리의 수가 아닌 경우 텍스트 `3자리의 수를 입력해 주세요.`
  - [x] 각 자릿수가 1에서 9사이의 숫자가 아닌 경우 텍스트 `각 자릿수는 1에서 9사이의 숫자로 입력해 주세요.`
  - [x] 각 자릿수가 각각 서로 다른 숫자가 아닌 경우 텍스트 `각 자릿수는 서로 다른 숫자로 입력해 주세요.`
  - [x] 입력이 1과 2 중 하나의 수가 아닌 경우 텍스트 `1과 2 중 하나로 입력해 주세요.`
  - [x] 게임 종료 텍스트 `게임이 종료됩니다.`

### 게임 로직 관련 (규칙)

- [x] 컴퓨터의 수 생성 (1에서 9까지 서로 다른 임의의 수 3개)
- [x] 볼, 스트라이크 개수 카운팅(return [볼 개수, 스트라이크 개수])
- [x] 낫싱 여부 판단 (0 === 볼 개수 === 스트라이크 개수)
- [x] 게임 클리어 여부 판단 (3스트라이크)
- [x] 클리어 할 때까지 로직 반복

## 개선이 필요한 부분

### 예외 처리 시 유저에게 이유 알려주기

![예외 처리 이유](https://user-images.githubusercontent.com/33623078/200576683-ddefabc5-174b-4e2d-890e-cf1c3494d6f0.gif)

현재 예외 처리 시 이유를 유저에게 알려주지 않는다.

#### validation.js

```js
const validateNumberInput = (userInputNumber) => {
  if (!isThreeDigitNumber(userInputNumber)) {
    throw threeDigitNumberValidationMessage;
  }
  ...
};
```

#### App.js

```js
try {
  const app = new App();
  app.play();
} catch (errorMessage) {
  errorMessage();
}
```

다음과 같이 `throw` 하는 에러 메시지를 받아 실행하고 싶지만 어떤 이유에서인지 동작하지 않는다.

함수가 같은 모듈에 있지 않아서 그런건지..? 원인 분석이 필요하다.

### 메시지 상수화

`숫자를 입력해주세요 : `

`게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.`

위와 같은 텍스트를 따로 분리하여 정리하는 것이 좋을 것 같다.

매번 작성하려니 번거롭고, 테스트 시 오타로 인해 통과가 되지 않을 수 있다.

### 테스트 작성 (jest)

[게임 실행 흐름](#기능-목록)의 항목들에 대해 테스트 코드를 작성했다.

하지만 jest가 아직은 익숙하지 못해 다음 항목들에 대해서는 작성하지 못했다.

- [x] 게임 반복
  - [x] 숫자 입력 요청 텍스트 `숫자를 입력해주세요 :` 출력 시 사용자 입력 받기
  - [x] 정답 시 게임 재시작 텍스트 `게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.` 출력 후 사용자 입력 받기
- [x] 게임 종료
  - [x] 입력이 유효하지 않은 경우

## 배운 점

- 커밋 메시지 규칙
- [Javscript 코드 컨벤션](https://github.com/airbnb/javascript)
- 함수 분리 (최대한 작게!)
- 테스트 코드 작성
  - `describe`, `it`
  - `jest.fn()`: 함수를 가짜 함수(mock functiton)로 만들어서 return값을 지정할 수 있다
  - `jest.spyOn()`: 함수의 구현을 가짜로 대체하지 않고, 해당 함수의 호출 여부와 어떻게 호출되었는지만을 알아낸다
  - [여러 expect 메소드](https://jestjs.io/docs/expect)
